Add a check for changes in study session message before reposting

Fox-Islam opened the pull request:
If people want to have notifications on for the events channel it might get annoying having the ping go off even if there are no changes

Fox-Islam committed:
Add a check for changes in study session message before reposting

If people want to have notifications on for the events channel
it might get annoying having the ping go off even if there are no
changes

chris-mcdonald-dev commented on src/tasks/studySession/channelReminder.js:
```diff
@@ -29,12 +29,16 @@ export default function sendChannelReminder(client) {
-            mostRecentStudySessionMessage.delete().then(() => {
-                makeStudySessionMessage().then((studyMessage) => {
-                    studySessionChannel.send(studyMessage).catch((error) => {
+            makeStudySessionMessage().then((studyMessage) => {
+                if (upcomingSessionsAreDifferent(mostRecentStudySessionMessage, studyMessage)) {
+                    mostRecentStudySessionMessage.delete().then(() => {
+                        studySessionChannel.send(studyMessage).catch((error) => {
+                            console.log(error);
+                        });
+                    }).catch((error) => {
                         console.log(error);
                     });
-                });
+                }
             }).catch((error) => {
```
With our current logic, it seems like we might be making a query to the database every minute after the 5 hours have passed until there is a change in session messages.
What do you think is the best way to handle this? I thought about changing the length of the interval, but that seems like it would get complicated.

Fox-Islam replied:
Ooh, yeah, you're right, I didn't notice that! I think there are two simple fixes:
1. Have an additional `cron.schedule` in scheduler.js that runs every five hours, and put this task in that instead of in the minute one 
2. Instead of checking if the post is 5 hours or more ago, check if the time between the post and now is exactly a multiple of 5 hours

Are either of these ok? Which one would you prefer? 

Edit: I've created an additional commit with the cron job change

chris-mcdonald-dev replied:
Smart! I like it a lot.

chris-mcdonald-dev reviewed (changes requested):
Our server is currently set to have notifications off server-wide by default unless users intentionally turn them on. Are you still getting notifications?
![image](https://user-images.githubusercontent.com/69683963/120733445-314a2700-c4b5-11eb-816e-de9d4e9b45c5.png)

Fox-Islam committed:
Create additional scheduler to handle channel reminder

Explicitly running the reminder task every 5 hours instead of
running it every minute and trying to handle the timing from
within the task logic

chris-mcdonald-dev commented on src/tasks/index.js:
```diff
@@ -1,3 +1,12 @@
+const notifySubscribers = studySessionTasks[0];
+const sendChannelReminder = studySessionTasks[1];
+
+const tasks = {
+    "minute": [],
+    "hour": [notifySubscribers],
+    "five-hour": [sendChannelReminder]
+};
```
I love the separation here! It definitely gives us a lot more control. What do you think about having this `sendChannelReminder` function get called when new study sessions/events are made?
Hopefully that would also remove any frustration of users scheduling a session and having to wait a number of hours before it shows up in the events channel.
https://github.com/chris-mcdonald-dev/korean-discord-app/blob/1741c8285d9d1d210a0e1af86b4b681d683c495d/src/scripts/activities/study-session.js#L53

Fox-Islam replied:
That makes a lot of sense! 

In fact, since that would handle all cases where a study session is added to the list (there's no other way to create a study session), the only thing that would be handled by the scheduler is the case when removing a study session from the list (either through cancellation or the date passing), which I think we can get away with being less urgent with an update for - could maybe drop the scheduled task time to 12-hourly? Daily?

Also, I haven't checked but I think once a study session passes it remains in the database forever, do you think those are useful to keep around? If not, I think it would be a "good first issue" to create a weekly/fortnightly/monthly scheduled task that deletes all study sessions that have passed

Fox-Islam committed:
Use session title in embed title

Fox-Islam committed:
Update upcoming sessions post whenever a session is created

Fox-Islam committed:
Increase time between scheduled event reminders to once per day
